projects/gloo/pkg/plugins/aws/ec2/ec2_utils.go

```
@@ -0,0 +1,119 @@
+package ec2
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/solo-io/go-utils/contextutils"
+	"go.uber.org/zap"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/gloo/projects/gloo/pkg/api/v1/plugins/aws/glooec2"
+	aws2 "github.com/solo-io/gloo/projects/gloo/pkg/utils/aws"
+)
+
+func GetEc2Session(ec2Upstream *glooec2.UpstreamSpec, secrets v1.SecretList) (*session.Session, error) {
+	return aws2.GetAwsSession(ec2Upstream.SecretRef, secrets, &aws.Config{Region: aws.String(ec2Upstream.Region)})
+}
+func ListEc2InstancesForCredentials(ctx context.Context, sess *session.Session, ec2Upstream *glooec2.UpstreamSpec) ([]*ec2.Instance, error) {
+	svc := ec2.New(sess)
+	contextutils.LoggerFrom(ctx).Debugw("ec2Upstream", zap.Any("spec", ec2Upstream))
+	input := &ec2.DescribeInstancesInput{
+		Filters: convertFiltersFromSpec(ec2Upstream),
+	}
+	contextutils.LoggerFrom(ctx).Debugw("ec2Upstream input", zap.Any("value", input))
+	result, err := svc.DescribeInstances(input)
+	contextutils.LoggerFrom(ctx).Debugw("ec2Upstream result", zap.Any("value", result))
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			// TODO - handle specific aws error codes
+			default:
+				contextutils.LoggerFrom(ctx).Errorw("unable to describe instances, aws error", zap.Error(aerr))
+			}
+		} else {
+			contextutils.LoggerFrom(ctx).Errorw("unable to describe instances, other error", zap.Error(err))
+		}
+	}
```

we should return the error here, else `getInstancesFromDescription` will cause a nil pointer panic